lms/djangoapps/lti_provider/outcomes.py

```diff
@@ -0,0 +1,199 @@
+        # service but did not have the consumer instance guid available, in
+        # which case the outcome service would only have the consumer key
+        # available. If we now have better information, we can improve that
+        # record.
+        if instance_guid and not outcomes.instance_guid:
+            outcomes.instance_guid = instance_guid
+            outcomes.save()
+
+        # Create a record for this assignment. Note that there may already
+        # be a record, if this assignment has been launched more than once
+        # by the same user/tool consumer combination.
+        GradedAssignment.objects.get_or_create(
+            lis_result_sourcedid=result_id,
+            course_key=course_key,
+            usage_key=usage_key,
+            user=user,
+            outcome_service=outcomes
+        )
+
+
+# Pylint doesn't recognize members in the LXML module
+# pylint: disable=no-member
+def generate_replace_result_xml(result_sourcedid, score):
+    """
+    Create the XML document that contains the new score to be sent to the LTI
+    consumer. The format of this message is defined in the LTI 1.1 spec.
+    """
+    envelope = etree.Element(
+        'imsx_POXEnvelopeRequest',
+        xmlns='http://www.imsglobal.org/services/ltiv1p1/xsd/imsoms_v1p0'
+    )
+    header = etree.SubElement(envelope, 'imsx_POXHeader')
+    header_info = etree.SubElement(header, 'imsx_POXRequestHeaderInfo')
+    version = etree.SubElement(header_info, 'imsx_version')
+    message_id = etree.SubElement(header_info, 'imsx_messageIdentifier')
+    body = etree.SubElement(envelope, 'imsx_POXBody')
+    replace_result_request = etree.SubElement(body, 'replaceResultRequest')
+    result_record = etree.SubElement(replace_result_request, 'resultRecord')
+    sourced_guid = etree.SubElement(result_record, 'sourcedGUID')
+    sourced_id = etree.SubElement(sourced_guid, 'sourcedId')
+    result = etree.SubElement(result_record, 'result')
+    result_score = etree.SubElement(result, 'resultScore')
+    language = etree.SubElement(result_score, 'language')
+    text_string = etree.SubElement(result_score, 'textString')
+
+    version.text = 'V1.0'
+    message_id.text = str(uuid.uuid4())
+    sourced_id.text = result_sourcedid
+    language.text = 'en'
+    text_string.text = str(score)
+
+    return etree.tostring(envelope, xml_declaration=True, encoding='UTF-8')
+
+
+def sign_and_send_replace_result(assignment, xml):
+    """
+    Take the XML document generated in generate_replace_result_xml, and sign it
+    with the consumer key and secret assigned to the consumer. Send the signed
+    message to the LTI consumer.
+    """
+    outcome_service = assignment.outcome_service
+    assert outcome_service, 'assignment.outcome_service guaranteed to be non-null'
+    consumer_key = outcome_service.consumer_key
+    assert consumer_key, 'outcome_service.consumer_key guaranteed to be non-null'
```

Comment:
I've always seen asserts as more documentation than error handling, with the premise that when you assert something it's because you absolutely know that it's true, and if it's not it's because some other developer (who might be a future version of yourself) messed up. In these cases, the asserts refer to checks that are made elsewhere in the code - I wouldn't expect anybody to catch an exception thrown from here, so it seems better not to throw one.